src/components/application_manager/src/state_controller_impl.cc

```diff
@@ -245,129 +245,185 @@ void StateControllerImpl::SetRegularState(ApplicationSharedPtr app,
+  const bool both_navi = app1.is_navi() && app2.is_navi();
+  const bool both_vc = app1.is_voice_communication_supported() &&
+                       app2.is_voice_communication_supported();
+  const bool both_media = (app1.is_media_application() && !app1.is_navi() &&
+                           !app1.is_voice_communication_supported()) &&
+                          (app2.is_media_application() && !app2.is_navi() &&
+                           !app2.is_voice_communication_supported());
+  const bool both_other =
+      !app1.IsAudioApplication() && !app2.IsAudioApplication();
+
+  const bool same_app_audio_type =
+      both_navi || both_vc || both_media || both_other;
+
+  return same_app_audio_type;
+}
 }  // unnamed namespace
 
 void StateControllerImpl::HmiLevelConflictResolver::operator()(
     ApplicationSharedPtr app_to_resolve) {
   DCHECK_OR_RETURN_VOID(app_to_resolve);
   DCHECK_OR_RETURN_VOID(state_ctrl_);
   DCHECK_OR_RETURN_VOID(applied_);
+  DCHECK_OR_RETURN_VOID(state_);
 
   if (app_to_resolve == applied_) {
     return;
   }
 
+  const HmiStatePtr state_to_resolve = app_to_resolve->RegularHmiState();
+  DCHECK_OR_RETURN_VOID(state_to_resolve);
+
   // Just for debug
-  //{
-  //  const std::string nl_indent = "\n ---> ";
-  //  std::stringstream ss;
-  //  ss << '\n';
-  //  ss <<
-  //  "-------------------HMI_CONFLICT_RESOLVER_IN_DATA-------------------";
-  //  ss << nl_indent << "Applied app id    - " << applied_->app_id();
-  //  ss << nl_indent << "Applied cur state - " << *applied_->CurrentHmiState();
-  //  ss << nl_indent << "Applied reg state - " << *applied_->RegularHmiState();
-  //  ss << nl_indent << "Resolve app id    - " << app_to_resolve->app_id();
-  //  ss << nl_indent << "Resolve cur state - "
-  //     << *app_to_resolve->CurrentHmiState();
-  //  ss << nl_indent << "Resolve reg state - "
-  //     << *app_to_resolve->RegularHmiState();
-  //  LOG4CXX_DEBUG(logger_, ss.str());
-  //}
-
-  // TODO(EKuliiev@luxoft.com) update logic description
-  // If applied Hmi state is FULL:
-  //  all not audio applications will get BACKGROUND
-  //  all applications with same HMI type will get BACKGROUND
-  //  all audio applications with other HMI type(navi, vc, media) in FULL will
-  //  get LIMMITED HMI level
-
-  // If applied Hmi state is LIMITED:
-  //  all applications with other HMI types will save HMI states
-  //  all not audio  applications will save HMI states
-  //  all applications with same HMI type will get BACKGROUND
-
-  // If applied Hmi state is BACKGROUND:
-  //  all applications will save HMI states
-
-  const HmiStatePtr reg_state_to_resolve = app_to_resolve->RegularHmiState();
-  DCHECK_OR_RETURN_VOID(reg_state_to_resolve);
-
-  // for applied app constants
-  const bool applied_grabs_full =
-      state_->hmi_level() == mobile_apis::HMILevel::HMI_FULL;
-
-  const bool applied_grabs_audio =
-      HasStreamingPermissions(state_->hmi_level()) &&
-      applied_->IsAudioApplication();
+  // {
```

Comment:
@EKuliiev This changes was moved to doxygen? Please try to avoid commits that override each other in one PR